### PR TITLE
Convert IIFE Arguments to Local Variables

### LINKED ISSUE #2100: Dead code elimination does not handle simple case
Consider simple code:

``` javascript
"use strict";
(function() {
    var module$library = {};
    var module$main = {};

    (function(exports) {
        function foo() {
            return "foo";
        }

        function bar() {
            return "bar";
        }

        exports.foo = foo;
        exports.bar = bar;
    })(module$library);

    (function(exports, library) {
        function baz() {
            console.log(library.foo());
        }
        baz();
        exports.baz = baz;
    })(module$main, module$library);
})();
```

GCC with advanced mode enables produces the following code:

``` javascript
(function(){var a={};(function(b){b.c=function(){return"foo"};b.a=function(){return"bar"}})(a);(function(b){function a(){console.log("foo")}a();b.b=a})({},a)})();
```

It's expected that functions `foo` and `bar` are eliminated. However, they don't. Actually, this code can be simplified to the form:

``` javascript
console.log("foo");
```

### PR DESCRIPTION
**Pull request by ChadKillingsworth**
Passing arguments to an IIFE seems to block a lot of future optimizations - most importantly function inlining. I believe in this case it's safe to convert the function args to local variable assignments.

Fixes #2100

### CODE REVIEW

**Comment by dimvar** [comment]
I'll take a look. Probably tomorrow.

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ * @author chadkillingsworth@gmail.com (Chad Killingsworth)
+ */
+public class ConvertIIFEArgsToVars extends NodeTraversal.AbstractPreOrderCallback
+    implements CompilerPass {
+  private final AbstractCompiler compiler;
+  private final List<Node> validIifes;
+
+  public ConvertIIFEArgsToVars(AbstractCompiler compiler) {
+    this.compiler = compiler;
+    validIifes = new ArrayList();
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: validIifes is unused.

File: test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            "  console.log(c, d);",
+            "})();"),
+        LINE_JOINER.join(
+            "(function() {",
+            "  var c = undefined;",
+            "  var d = undefined;",
+            "  console.log(c, d);",
+            "})();"));
+  }
+}
```
Comment (dimvar) [inline test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java]: Can you add a test where the arguments are more than the formal parameters? I wanna make sure we don't crash in that case.

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  @Override
+  public void process(Node externs, Node root) {
+   NodeTraversal.traverseEs6(compiler, root, this);
+  }
+
+  public boolean shouldTraverse(NodeTraversal t, Node n, Node parent) {
+    return true;
+  }
+
+  public void visit(NodeTraversal t, Node n, Node parent) {
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: This doesn't check that the IIFE is an expr-result. Does that matter, or do you want to do this optimization for all iifes?

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
+/*
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: I'm having some trouble with the internal tools while importing the PR to test it.

I've left a few initial comments below.

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java (hunk 1 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: removed

File: test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
+package com.google.javascript.jscomp;
+
+/**
+ * Tests for {@link ConvertIIFEArgsToVars}.
+ */
+public final class ConvertIIFEArgsToVarsTest extends CompilerTestCase {
```
Comment (dimvar) [inline test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java]: Add test about variable capture

 public void testPreventAccidentalVariableCapture() {
    test(
        LINE_JOINER.join(
            "var a = 1;",
            "(function(c) {",
            "  var a = 2;",
            "  console.log(c);",
            "})(a)"),
        LINE_JOINER.join(
            "var a = 1;",
            "(function() {",
            "  var c = a;",
            "  var a$jscomp$1 = 2;",
            "  console.log(c);",
            "})()"));
  }

File: test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
+package com.google.javascript.jscomp;
```
Comment (dimvar) [inline test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java]: Add license

File: src/com/google/javascript/jscomp/DefaultPassConfig.java [hunk 7]
```diff
@@ -548,6 +548,12 @@ private void addNonTranspilationCheckPasses(List<PassFactory> checks) {
       passes.add(runtimeTypeCheck);
     }
 
+    // Converts arguments to IIFEs to variable assignments within the function.
+    // Improves later optimizations such as function inlining.
+    if (options.optimizeCalls && options.inlineLocalVariables) {
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/DefaultPassConfig.java]: Why did you guard with two options here instead of just optimizeCalls?

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      newVars.add(IR.var(param.cloneNode(), getCallArgument(param, call))
+          .useSourceInfoFromForTree(param));
+
+      param = param.getNext();
+    }
+    funcParams.removeChildren();
+    for (int i = newVars.size() - 1; i >= 0; i--) {
+      funcParams.getNext().addChildToFront(newVars.get(i));
+    }
+
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: Add comment here:

// Remove all arguments of the call node, except "this" if it exists.

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        n.getNext() != null && n.getNext().isString() &&
+        n.getNext().getString().equals("call") &&
+        parent.getParent() != null && parent.getParent().isCall() &&
+        parent.getParent().getFirstChild() == parent) {
+      call = parent.getParent();
+    } else {
+      return;
+    }
+
+    Node funcParams = n.getSecondChild();
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: Use NodeUtil.getFunctionParameters here, it's easier to understand

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  @Override
+  public void visit(NodeTraversal t, Node n, Node parent) {
+    if (!n.isFunction() || parent == null) {
+      return;
+    }
+
+    Node call;
+
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: Define grandparent here to clean up the condition of the else-if.

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+import com.google.javascript.rhino.IR;
+import com.google.javascript.rhino.Node;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Convert the arguments of an immediately invoked function expression to variables
+ * and assign the value from the call arguments. This improves function inlining.
+ *
+ * Since passes which implement OptimizeCalls only handle named references, this is it's own pass.
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: What do you mean by "named referenes" here? Functions with a name? Can you rephrase it?

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+package com.google.javascript.jscomp;
+
+import com.google.common.base.Preconditions;
+import com.google.javascript.rhino.IR;
+import com.google.javascript.rhino.Node;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Convert the arguments of an immediately invoked function expression to variables
```
Comment (dimvar) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: Add to the description:

This pass relies on variable renaming (done during normalization) to avoid accidental variable capture.

Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: But does it? In all the cases I've come up with (including the test you suggested above), the variable was already shadowed by either the parameter name or by an existing local variable:

```js
var a = 1;
(function(a) { // shadows a
  console.log(a);
})(1)

(function(b) { 
  var a = b; // shadows a
  console.log(a);
})(1)
```

Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: Nevermind - I see what you are saying now. I'll update the description.

File: src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java (hunk 11 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java]: Yes done.

File: test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java (hunk 5 above)
Comment (ChadKillingsworth) [inline test/com/google/javascript/jscomp/ConvertIIFEArgsToVarsTest.java]: Good call - added.

**Comment by dimvar** [comment]
LGTM. I will submit this internally. Thanks Chad.

**Comment by concavelenz** [comment]
I don't understand why this passes was added why not instead adjust the inlining heurstic which is deciding not to inline these functions (likely because it needs to introduce variables)?

**Comment by ChadKillingsworth** [comment]
A lot of the precursor work to inlining didn't run on IIFEs because it only worked off of named functions (Optimize calls, optimize params, etc). This was an easy pass for me to write without delving into inlining mechanics.

It's a real world problem though - people love to pass random objects through IIFE arguments.

**Comment by concavelenz** [comment]
optimize calls, optimize params were design to benefit functions that could not be inlined (no call site uses the return value, every call passes the same constant, etc) and aren't "precursors" to inlining.    If a function is inlinable after a parameter has been removed from a function due to optimize parameters or after the return value has be removed because no call site uses it, that is because the heuristic has decided it is now cheaper to inline.  Generally, the function inliner is more aggressive with IIFE (because they are only ever called once) but it might be reasonable to "always inline IIFEs" (assuming they don't use arguments or eval or one of the other constructs that the inliner bails on).

**Comment by Dominator008** [comment]
@concavelenz I think we tried "always inline IIFE" but it was rolled back:

https://github.com/google/closure-compiler/commit/0660acdb4691cade499563c663a2c83243d0c4b2

**Comment by concavelenz** [comment]
Right, that is just one possible heuristic change.  It sounds like @ChadKillingsworth just want the inliner to be more aggressive and I don't see that we need an additional pass for that.

**Comment by ChadKillingsworth** [comment]
I think like most things, we just need someone to A) suggest another way to solve the problem and B) do the work to switch it. This was just one of those interesting issues I came across and solved because I thought of a way to do it.

Original use case:
```js
"use strict";
(function() {
    var module$library = {};
    var module$main = {};

    (function(exports) {
        function foo() {
            return "foo";
        }

        function bar() {
            return "bar";
        }

        exports.foo = foo;
        exports.bar = bar;
    })(module$library);

    (function(exports, library) {
        function baz() {
            console.log(library.foo());
        }
        baz();
        exports.baz = baz;
    })(module$main, module$library);
})();
```

Can the inliner handle property assignments like that?

**Comment by MatrixFrog** [comment]
It looks like this causes some compiler crashes. Details shortly.

**Comment by MatrixFrog** [comment]
Minimal repro is `(function(f) {})(function() {});`

It looks like the 2nd function gets detached but then we still to traverse it, and `NodeUtil.isStatement` fails because it doesn't have a parent at that point.

```
null
  Node(FUNCTION ): input0:3:3
})(function() {});
  Parent: NULL
	at com.google.common.base.Preconditions.checkState(Preconditions.java:431)
	at com.google.javascript.jscomp.NodeUtil.isStatementParent(NodeUtil.java:2572)
	at com.google.javascript.jscomp.NodeUtil.isStatement(NodeUtil.java:2558)
	at com.google.javascript.jscomp.NodeUtil.isFunctionExpression(NodeUtil.java:2847)
	at com.google.javascript.jscomp.NodeTraversal.traverseFunction(NodeTraversal.java:653)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:634)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:718)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:642)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:718)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:642)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:718)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:642)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:718)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:642)
	at com.google.javascript.jscomp.NodeTraversal.traverse(NodeTraversal.java:330)
	at com.google.javascript.jscomp.NodeTraversal.traverseEs6(NodeTraversal.java:592)
	at com.google.javascript.jscomp.ConvertIIFEArgsToVars.process(ConvertIIFEArgsToVars.java:46)
```

**Comment by dimvar** [comment]
I'll take a look at the crash.

**Comment by concavelenz** [comment]
What the inliner does is basically what you do here: it introduces a variable to represent the parameter and is fully capable of inlining the function IF it decides to.  It is just a question of telling it to.

**Comment by dimvar** [comment]
As mentioned in the internal thread, I tried inlining IIFEs in the past and did it incorrectly and rolled it back. The inliner is pretty complex and I don't have time at the moment to devote to it. You know the inliner better than me. If you want to improve it for this case, to handle Chad's tests, it's totally fine by me to delete this pass. I just don't have time to look into this myself.

**Comment by concavelenz** [comment]
Yes, inlining isn't simple but that doesn't mean that we should be writing code to work around it.  Here, as far I understand it, we are just trying to encourage inlining and perhaps by pass that "assumeClosuresOnlyCaptureReferences" is false by default.  (CompilationLevel also sets "options.setAssumeClosuresOnlyCaptureReferences(false);" explicitly because it was true by default at one time.

**Comment by ChadKillingsworth** [comment]
`setAssumeClosuresOnlyCaptureReferences(true)` does inline the sample code. Trying to track down if we removed that restriction from IIFEs if the problems originally reported would re-occur.

**Comment by concavelenz** [comment]
I believe that Chad's pass makes that same assumptions as setAssumeClosuresOnlyCaptureReferences(true) except it restricts it to IIFEs only.  As per off-line discussion Chad was going to attempt to enable this logic for IIFEs only and remove the pass.

**Comment by dimvar** [comment]
Great!

**Comment by MatrixFrog** [comment]
It looks like we found one more issue with this change: https://closure-compiler-debugger.appspot.com/#input0%3D(function(window)%2520%257B%250A%2520%2520alert(window)%253B%250A%257D)(this)%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26MISSING_PROPERTIES%3D1%26OPTIMIZE_CALLS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

If there is a `'use strict'` directive this will make `window` undefined.

**Comment by ChadKillingsworth** [comment]
I'm guessing we just need to not optimize an IIFE that has `this` in the arguments.

**Comment by MatrixFrog** [comment]
Or `arguments`: https://closure-compiler-debugger.appspot.com/#input0%3D(function(_)%2520%257B%250A%2520%2520(function(args)%2520%257B%250A%2520%2520%2520%2520console.log(args%255B0%255D)%253B%250A%2520%2520%257D)(arguments)%253B%250A%257D)(0)%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26MISSING_PROPERTIES%3D1%26OPTIMIZE_CALLS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

@Dominator008 found this comment which may be helpful: https://github.com/google/closure-compiler/blob/55cf43ee31e80d89d7087af65b5542aa63987874/src/com/google/javascript/jscomp/OptimizeParameters.java#L353-L355

**Comment by ChadKillingsworth** [comment]
I already bail on `arguments`: https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/ConvertIIFEArgsToVars.java#L102

**Comment by ChadKillingsworth** [comment]
Ahh but not as a call I argument. Fixing ...

**Comment by concavelenz** [comment]
I wonder if we should just try the change to the InlineFunctions again?  The function inliner handles all this stuff properly.
